### PR TITLE
Refactor bindings, introducing joinBindings

### PR DESCRIPTION
**Pull request by LPGhatguy**
Inspired by #204. 😄 

Reworks bindings from the ground up and adds support for joining bindings together.

Checklist before submitting:
* [x] Added entry to `CHANGELOG.md`
* [x] Added/updated relevant tests
* [x] Added/updated documentation

### CODE REVIEW

File: src/Binding.lua [hunk 1]
```diff
@@ -4,146 +4,156 @@ local Type = require(script.Parent.Type)
...
+	function impl.subscribe(callback)
+		return BindingInternalApi.subscribe(upstreamBinding, function(newValue)
+			callback(predicate(newValue))
 		end)
 	end
 
-	local disconnect = internalData.changeSignal:subscribe(handler)
-	internalData.subscriberCount = internalData.subscriberCount + 1
+	function impl.update(newValue)
+		-- This operation doesn't make sense; we can leave it as a no-op
```
Comment (AmaranthineCodices) [inline src/Binding.lua]: Let's throw here to catch weird cases (like using a mapped binding as a ref)

File: src/Binding.lua [hunk 2]
```diff
@@ -4,146 +4,156 @@ local Type = require(script.Parent.Type)
...
+end
+
+function BindingInternalApi.join(upstreamBindings)
+	if config.typeChecks then
+		assert(typeof(upstreamBindings) == "table", "Expected arg #1 to be of type table")
+
+		for key, value in pairs(upstreamBindings) do
+			if Type.of(value) ~= Type.Binding then
+				local message = (
+					"Expected arg #1 to contain only bindings, but key %s had a non-binding value"
```
Comment (AmaranthineCodices) [inline src/Binding.lua]: Should we use `%q` here, in case the key has spaces or something similarly odd?

File: src/Binding.lua [hunk 3]
```diff
@@ -4,146 +4,156 @@ local Type = require(script.Parent.Type)
...
 
-	local setter = function(newValue)
-		Binding.update(binding, newValue)
+			disconnects = nil
+		end
 	end
 
-	return binding, setter
+	function impl.update(newValue)
+		-- This operation doesn't make sense; we can leave it as a no-op
```
Comment (AmaranthineCodices) [inline src/Binding.lua]: Should throw here too.

File: src/Binding.lua [hunk 4]
```diff
@@ -4,146 +4,156 @@ local Type = require(script.Parent.Type)
...
---[[
-	Default mapping function used for non-mapped bindings
-]]
-local function identity(value)
-	return value
+local BindingImpl = Symbol.named("BindingImpl")
+
+local BindingInternalApi = {}
+
+local BindingPublicMethods = {}
```
Comment (Reselim) [inline src/Binding.lua]: Should this be changed to `bindingPrototype` to be congruent with other classes?

File: src/Binding.lua [hunk 5]
```diff
@@ -4,146 +4,156 @@ local Type = require(script.Parent.Type)
...
-end
 
---[[
-	Create a new binding object with the given starting value. This
-	function will be exposed to users of Roact.
-]]
-function Binding.create(initialValue)
-	local binding = {
-		[Type] = Type.Binding,
+	function impl.subscribe(callback)
```
Comment (Reselim) [inline src/Binding.lua]: Would it make more sense to only subscribe to `upstreamBindings` once and create/destroy the upstream connections depending on if there are subscribers or not? The prior implementation does this.

Upstream connections could easily add up, considering bindings mapped from this one will invoke `subscribe` when subscribed to. A good example of this potentially becoming a problem would be when animating transparency.

File: src/Binding.lua [hunk 6]
```diff
@@ -4,146 +4,156 @@ local Type = require(script.Parent.Type)
...
-
-		--[[
-			If our subscribers count drops to 0, we can safely unsubscribe from
-			our source binding
-		]]
-		if internalData.subscriberCount == 0 and internalData.upstreamDisconnect ~= nil then
-			internalData.upstreamDisconnect()
-			internalData.upstreamDisconnect = nil
+	local impl = {
+		disconnectMethods = nil,
```
Comment (Reselim) [inline src/Binding.lua]: This variable is unused.

File: src/Binding.lua (hunk 5 above)
Comment (LPGhatguy) [reply src/Binding.lua]: Just got back into working on this!

I don't think it's a big deal to pass subscriptions up through to their parents. At the end of the day, someone is going to have to track the connection, and consolidating that at the top is the simplest approach!

**Comment by LPGhatguy** [comment]
PR feedback was addressed and I added a changelog entry and (bad) documentation.

I'm not sure what a small sample for how to use `joinBindings` would be; the current one is quite large.

**Comment by Reselim** [comment]
LGTM! Works well in my (rather large) UI codebase making use of `joinBindings`.

**Comment by LPGhatguy** [comment]
Will merge this first thing Monday to cut 1.1!

**Pull request merged**